crates/eframe/src/web/web_painter_wgpu.rs

```diff
@@ -83,46 +83,95 @@ impl WebPainterWgpu {
         })
     }
 
-    #[allow(unused)] // only used if `wgpu` is the only active feature.
-    pub async fn new(canvas_id: &str, options: &WebOptions) -> Result<Self, String> {
-        log::debug!("Creating wgpu painter");
-
-        {
+    pub async fn init_wgpu_and_create_render_state(
+        mut backends: wgpu::Backends,
+        wgpu_options: &egui_wgpu::WgpuConfiguration,
+        depth_format: Option<wgpu::TextureFormat>,
+        canvas: &HtmlCanvasElement,
+    ) -> Result<(RenderState, wgpu::Surface<'static>), String> {
+        // Don't try WebGPU if we're not in a secure context.
+        if backends.contains(wgpu::Backends::BROWSER_WEBGPU) {
             let is_secure_context = web_sys::window().map_or(false, |w| w.is_secure_context());
             if !is_secure_context {
                 log::info!(
-                    "WebGPU is only available in secure contexts, i.e. on HTTPS and on localhost"
+                    "WebGPU is only available in secure contexts, i.e. on HTTPS and on localhost."
                 );
+
+                // Don't try WebGPU since we established now that it will fail.
+                backends.remove(wgpu::Backends::BROWSER_WEBGPU);
+
+                if backends.is_empty() {
+                    return Err("No available supported graphics backends.".to_owned());
+                }
             }
         }
 
         let instance = wgpu::Instance::new(wgpu::InstanceDescriptor {
-            backends: options.wgpu_options.supported_backends,
+            backends,
             ..Default::default()
         });
-
-        let canvas = super::canvas_element_or_die(canvas_id);
-
         let surface = instance
             .create_surface(wgpu::SurfaceTarget::Canvas(canvas.clone()))
             .map_err(|err| format!("failed to create wgpu surface: {err}"))?;
+        let render_state_result =
+            RenderState::create(wgpu_options, &instance, &surface, depth_format, 1).await;
+
+        if let Err(egui_wgpu::WgpuError::NoSuitableAdapterFound) = &render_state_result {
+            if backends == wgpu::Backends::BROWSER_WEBGPU | wgpu::Backends::GL {
```

Comment:
nit: I think this logic is slightly clearer

```suggestion
            if backends.contains(Backends::BROWSER_WEBGPU) && backends.contains(Backends::GL) {
```

Reply:
... and might actually be necessary if somieone also sets other backends that are not for web!